doc/port_link_training/port-link-training-design.md

```diff
@@ -0,0 +1,363 @@
+
+#### YANG Model Enhancements
+
+The port yang model needs to update according to DB schema change. The yang model changes of new fields are described below:
+
+```
+leaf link_training {
+  type string {
+    pattern "on|off";
+  }
+}
+```
+
+These new yang leaf will be added to sonic-port.yang.
+
+#### DB Migration Considerations
+
+By having port link training disabled if "link_training" is not specified in the CONFIG_DB, this feature is fully backward compatible.
+
+#### SWSS Enhancements
+
+##### Port Link-Training Ability Flag
+
+During system startup, PortsOrch should query for the per-port link-training abilities from syncd, and populate these flags into **LINK_TRAINING** table of [STATE_DB](#state-db-enhancements)
+
+##### Port Configuration Flow
+
+The current SONiC port configuration flow in PortsOrch can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+```
+
+The new SONiC port configuration flow can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+
+if link_training changed:
+    setPortLinkTraining(port, link_training)
+```
+
+##### Link Training Status Refresh
+
+A timer will also be introduced into PortsOrch for polling link-training status:
+
+```
+while PortsOrch is alive:
+    sleep_sec(3)
+    for port in all_port_list:
+        updatePortLinkTrainingStatus(port)
+```
+
+The updatePortLinkTrainingStatus(port) can be furtherly described in following pseudo code:
+
+```
+if port is not physical_port:
+    return
+
+status = "off"
+if link-training is supported by the SAI and is enabled:
+    status = getLinkTrainingRxStatus(port)
+    if status != trained:
+        error = getLinkTrainingFailure(port)
+        if error != "none":
+            status = error
+        else:
+            status = "not_trained"
+updateLinkTrainingStateDB(port, status)
+```
+
+
+### Warmboot Design Impact
+
+SAI and lower layer must not flap port during warmboot no matter what link training parameter is given.
+
+### Limitations
+
+N/A
+
+### Testing Requirements
+
+#### Unit Test cases
+
+For **sonic-swss**, we will leverage the existing [test_port.py](https://github.com/Azure/sonic-swss/blob/master/tests/test_port.py) for this. A few new test cases will be added:
+
+1. Test attribute **link_training** on both direct and warm-reboot scenario. Verify SAI_PORT_ATTR_LINK_TRAINING_ENABLE is in ASIC_DB and has correct value.
+
+For **sonic-utilities**, we will leverage the existing [unit test framework](https://github.com/Azure/sonic-utilities/tree/master/tests) for this. A few new test cases will be added:
+
+1. Test command `config interface link-training <interface_name> <mode>`. Verify the command return error if given invalid interface_name or mode.
+
+#### System Test cases
+
+Will leverage sonic-mgmt to test this feature.
```

Comment:
can you be more specific?

Reply:
will there be new test cases added for this new feature?

Reply:
Thanks for the comment, and the HLD is now updated, please check it out and see if it works